### PR TITLE
feat: handle conda env ymls

### PR DESCRIPTION
**Pull request by mederka**
Here's what an example script with conda env yml could look like:
```
from isolate.backends.remote import IsolateServer

with open("env.yaml") as f:
    env_yml_str = f.read()

target_config = {
    "kind": "conda",
    "configuration": {
        "env_yml_str": env_yml_str
    }
}

config = {
    "host": "localhost:50001",
    "target_environments": [target_config]
}

env = IsolateServer.from_config(config)
```
Where `env.yaml` is:

```
name: testo
channels:
  - defaults
dependencies:
  - python=3.8.10
  - pip=20.3
  - pip:
    - isolate==0.7.1
    - dill==0.3.6
```

### CODE REVIEW

File: src/isolate/backends/conda.py [hunk 1]
```diff
@@ -87,28 +98,47 @@ def create(self) -> Path:
...
+                    f.write(self.env_yml_str)
+                try:
+                    self._run_conda(
+                        "env",
+                        "create",
+                        "-f",
+                        filename,
+                        "--prefix",
+                        env_path
+                    )
```
Comment (chamini2) [inline src/isolate/backends/conda.py:113]: no `dependencies = self._compute_dependencies()` for this case?

Comment (chamini2) [reply src/isolate/backends/conda.py:113]: We get the python version from it

File: src/isolate/backends/conda.py [hunk 2]
```diff
@@ -155,3 +185,8 @@ def _get_conda_executable() -> Path:
             "Could not find conda executable. If conda executable is not available by default, please point isolate "
             " to the path where conda binary is available 'ISOLATE_CONDA_HOME'."
         )
+
+
+def _is_isolate_in_yml(env_yml: str) -> bool:
+    # TODO: make this check more in-depth
+    return '- isolate==' in env_yml or '- isolate\n' in env_yml
```
Comment (chamini2) [inline src/isolate/backends/conda.py:192]: You can parse the yaml and look for it as a python object

File: src/isolate/backends/conda.py [hunk 3]
```diff
@@ -33,6 +34,14 @@ class CondaEnvironment(BaseEnvironment[Path]):
...
     packages: List[str] = field(default_factory=list)
     python_version: Optional[str] = None
+    env_yml_str: Optional[str] = None
+
+    def __post_init__(self):
+        # When using env_yml_str, the new conda environment will come without isolate,
+        # unless isolate is a dependency.
+        message = "isolate not found in environment YAML file. Please add isolate as a pip dependency."
+        if self.env_yml_str and not _is_isolate_in_yml(self.env_yml_str):
+            raise EnvironmentCreationError(message)
```
Comment (chamini2) [inline src/isolate/backends/conda.py:44]: what about offering a `env_dict` value

and if `env_yml_str` is present, we parse `env_yml_str` in the `env_dict`

**Review by chamini2 — COMMENTED** [review]
some questions

File: src/isolate/backends/conda.py [hunk 4]
```diff
@@ -33,6 +35,14 @@ class CondaEnvironment(BaseEnvironment[Path]):
 
     packages: List[str] = field(default_factory=list)
     python_version: Optional[str] = None
+    env_yml_str: Optional[str] = None
```
Comment (isidentical) [inline src/isolate/backends/conda.py:38]: Does it make sense to drop `env_yml_str` completely from here and only support `env_dict`? If the `env_yml_str` is passed during the `from_config` method, and if pyyaml is installed; we can parse it and convert.

E.g.:
```py
    @classmethod
    def from_config(
        cls,
        config: Dict[str, Any],
        settings: IsolateSettings = DEFAULT_SETTINGS,
    ) -> BaseEnvironment:
        env_yml_str = config.pop("env_yml_str", None)
        if env_yml_str:
            if "env_dict" in config: raise ...
            try:
                import yaml
            except ModuleNotFoundError:
                raise ... ("install pyyaml optional dependency")
            else:
                config["env_dict"] = yaml.safe_load(env_str)
                
        environment = cls(**config)
        environment.apply_settings(settings)
        return environment
```

File: src/isolate/backends/conda.py [hunk 5]
```diff
@@ -87,28 +105,48 @@ def create(self) -> Path:
...
-                    env_path,
-                    *dependencies,
-                )
-            except subprocess.SubprocessError as exc:
-                raise EnvironmentCreationError("Failure during 'conda create'") from exc
-
-        self.log(f"New environment cached at '{env_path}'")
-        return env_path
+            if self.env_dict:
+                filename = f"{str(uuid.uuid4())}.yaml"
```
Comment (isidentical) [inline src/isolate/backends/conda.py:109]: NIT: I think we can also use tempfile.NamedTemporaryFile/mkstemp as well for auto-cleaning

**Review by isidentical — APPROVED** [review]
This looks amazing @mederka, thanks so much for working on it! This also adds support for pyenv/conda to our internal isolate worker which even cooler (should we create another ticket to link to this PR as well!). I only had two minor suggestions regarding the configuration options, but we can merge this as is and work them in the second iteration as well if that would unblock you on the demo you've been working on.

File: src/isolate/backends/conda.py (hunk 5 above)
Comment (chamini2) [reply src/isolate/backends/conda.py]: This is to avoid having to comeup with a uuid? it takes care for us?

File: src/isolate/backends/conda.py [hunk 6]
```diff
@@ -87,28 +105,47 @@ def create(self) -> Path:
...
+                    yaml.dump(self.env_dict, f)
+                try:
+                    self._run_conda(
+                        "env",
+                        "create",
+                        "-f",
+                        filename,
+                        "--prefix",
+                        env_path
+                    )
```
Comment (chamini2) [inline src/isolate/backends/conda.py:122]: From what I'm seeing in [tempfile](https://github.com/python/cpython/blob/3.11/Lib/tempfile.py#L494-L509), I think you have to `with` the whole tempfile

```suggestion
            if self.env_dict:
                self.env_dict['dependencies'] = self._compute_dependencies()
                with tempfile.NamedTemporaryFile(suffix='.yml') as tf:
                  yaml.dump(self.env_dict, f)
                  try:
                      self._run_conda(
                          "env",
                          "create",
                          "-f",
                          filename,
                          "--prefix",
                          env_path
                      )
```

Comment (chamini2) [reply src/isolate/backends/conda.py:122]: this automatically `close`s the file which in turn deletes it for you

Comment (isidentical) [reply src/isolate/backends/conda.py:122]: Tip: sometimes tempfile buffer a bit on write, so let's also force flush it before conda env create call to ensure the contents are there before we run it

Comment (mederka) [reply src/isolate/backends/conda.py:120]: For some reason I couldn't get this to work. But doing it outside `with` clause works fine. 🤷

Comment (chamini2) [reply src/isolate/backends/conda.py:120]: Maybe you have to set mode so it is not the default? https://docs.python.org/3/library/tempfile.html#tempfile.TemporaryFile

```
with tempfile.NamedTemporaryFile(mode="w", suffix='.yml') as tf
```

**Pull request merged**